### PR TITLE
test(core): add incremental hydration tests

### PR DESCRIPTION
**Pull request by thePunderWoman**
This adds some additional tests for incremental hydration around registry and contract cleanup.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/defer/registry.ts [hunk 1]
```diff
@@ -49,6 +54,9 @@ export class DehydratedBlockRegistry {
       this.invokeTriggerCleanupFns(blockId);
       this.hydrating.delete(blockId);
     }
+    if (this.registry.size === 0) {
+      this.contract.instance?.cleanUp();
+    }
   }
 
   get size(): number {
```
Comment (AndrewKushnir) [inline packages/core/src/defer/registry.ts:62]: I guess we can remove this getter now (looks like it's unused now)?

File: packages/platform-server/test/incremental_hydration_spec.ts [hunk 2]
```diff
@@ -1514,5 +1520,233 @@ describe('platform-server partial hydration integration', () => {
...
+            }
+          </main>
+        `,
+      })
+      class SimpleComponent {
+        fnA() {}
+
+        registry: DehydratedBlockRegistry = inject(DEHYDRATED_BLOCK_REGISTRY);
+        jsActionMap: Map<string, Set<Element>> = inject(JSACTION_BLOCK_ELEMENT_MAP);
+        contract: EventContractDetails = inject(JSACTION_EVENT_CONTRACT);
```
Comment (AndrewKushnir) [inline packages/platform-server/test/incremental_hydration_spec.ts:1550]: nit: those types are likely inferred by TS:
```suggestion
        registry = inject(DEHYDRATED_BLOCK_REGISTRY);
        jsActionMap = inject(JSACTION_BLOCK_ELEMENT_MAP);
        contract = inject(JSACTION_EVENT_CONTRACT);
```

File: packages/platform-server/test/incremental_hydration_spec.ts [hunk 3]
```diff
@@ -1514,5 +1520,233 @@ describe('platform-server partial hydration integration', () => {
...
+      const nested = doc.getElementById('nested')!;
+      const clickEvent2 = new CustomEvent('click', {bubbles: true});
+      nested.dispatchEvent(clickEvent2);
+      await timeout(1000); // wait for defer blocks to resolve
+      appRef.tick();
+
+      expect(registry.size).toBe(0);
+      expect(jsActionMap.size).toBe(0);
+      expect(contract.instance!.cleanUp).toHaveBeenCalled();
+    }, 100_000);
```
Comment (AndrewKushnir) [inline packages/platform-server/test/incremental_hydration_spec.ts:1602]: Was that timeout helpful for testing/debugging only and can be removed?
```suggestion
    });
```

**Review by AndrewKushnir — APPROVED** [review]

File: packages/core/src/defer/registry.ts (hunk 1 above)
Comment (thePunderWoman) [reply packages/core/src/defer/registry.ts:62]: Actually it is in use in the `triggering.ts` file, which doesn't show in this diff. I've instead swapped line 57 to use this getter.

Comment (AndrewKushnir) [reply packages/core/src/defer/registry.ts:62]: I only see 1 instance where we use `size` in the mentioned file: https://github.com/angular/angular/blob/main/packages/core/src/defer/triggering.ts#L415 and you've removed the usage from there in this PR. Are there any other places?

Comment (thePunderWoman) [reply packages/core/src/defer/registry.ts:62]: Yeah, I corrected myself. The size getter is used extensively in the `incremental_hydration_spec.ts` file. You actually can see its usage in the diff, now that I look at it.

Comment (thePunderWoman) [reply packages/core/src/defer/registry.ts:62]: Actually, after a little more thought, it's only used in the new tests and doesn't make sense to keep around just for that. I'll get rid of it.

Comment (thePunderWoman) [reply packages/core/src/defer/registry.ts:62]: Crap...except nothing else exposes that information. I guess we do need it for now.

**Comment by thePunderWoman** [comment]
This PR was merged into the repository by commit e3343620b529bfe0e186c9292325b1158be89b58.

The changes were merged into the following branches: main, 19.0.x

**Pull request closed without merging**